lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp

```diff
@@ -589,9 +601,11 @@ Desugar(swift::Demangle::Demangler &dem, swift::Demangle::NodePointer node,
 
 /// Iteratively resolve all type aliases in \p node by looking up their
 /// desugared types in the debug info of module \p M.
-static swift::Demangle::NodePointer GetCanonicalNode(
-    TypeSystemSwiftTypeRef *module_holder, SwiftASTContext *target_holder,
-    swift::Demangle::Demangler &dem, swift::Demangle::NodePointer node) {
+static swift::Demangle::NodePointer
+GetCanonicalNode(TypeSystemSwiftTypeRef *module_holder,
+                 std::function<SwiftASTContext *()> target_holder_provider,
```

Comment:
IIRC, I renamed this to `target_holder` because in the end we wanted to read the Target* from it. If that's true, maybe this should be a std::function<Target *()> target_provider?

Reply:
I agree, that sounds better.

Reply:
Edit: std::function<TargetWP()>

Reply:
There's one situation where we need the `ASTContext`, so I think keeping it returning a SwiftASTContext is probably better.

```
  } Impl(swift_ast_context ? swift_ast_context->GetASTContext() : nullptr);
```

Reply:
That's in `GetClangTypeNode(CompilerType clang_type, swift::Demangle::Demangler &dem,  SwiftASTContext *swift_ast_context)` which doesn't call the argument `target_holder`. Maybe those two use-cases can be treated separately?

Reply:
Uh.. it's unfortunate that we need this there unconditionally. I guess the next task will be to see if we can factor out `AST->getSwiftName()` to not need a full SwiftASTContext initialization.

Reply:
I think it's better to keep it like this for now, as we're threading to both places from the same intermediary functions, and I think it'd be more confusing to treat them as separate (for now). If we factor out `AST->getSwiftName()` then we'll probably have to separate the cases.